core/client/fs/src/main/java/alluxio/client/block/AlluxioBlockStore.java

```diff
@@ -170,24 +169,22 @@ public BlockInStream getInStream(long blockId, InStreamOptions options,
       info = masterClientResource.get().getBlockInfo(blockId);
     }
     List<BlockLocation> locations = info.getLocations();
-    List<BlockWorkerInfo> blockWorkerInfo = Collections.EMPTY_LIST;
+    List<BlockWorkerInfo> blockWorkerInfo = getEligibleWorkers();
+    if (blockWorkerInfo.isEmpty()) {
+      throw new UnavailableException(ExceptionMessage.NO_WORKER_AVAILABLE.getMessage());
+    }
     // Initial target workers to read the block given the block locations.
     Set<WorkerNetAddress> workerPool;
     // Note that, it is possible that the blocks have been written as UFS blocks
     if (options.getStatus().isPersisted()
         || options.getStatus().getPersistenceState().equals("TO_BE_PERSISTED")) {
-      blockWorkerInfo = getEligibleWorkers();
       workerPool = blockWorkerInfo.stream().map(BlockWorkerInfo::getNetAddress).collect(toSet());
-      if (workerPool.isEmpty()) {
-        throw new UnavailableException(
-            "No Alluxio worker available. Check that your workers are still running");
-      }
     } else {
-      workerPool = locations.stream().map(BlockLocation::getWorkerAddress).collect(toSet());
-      if (workerPool.isEmpty()) {
-        throw new NotFoundException(
+      if (locations.isEmpty()) {
+        throw new UnavailableException(
```

Comment:
`getEligibleWorkers` could be expensive, so maybe we should only do this in the specific error cases? For example, here, we could check all the workers to see if any exist. So, it could look something like:

```java
if (locations.isEmpty()) {
  if (getEligibleWorkers().isEmpty()) {
    throw new UnavailableException(ExceptionMessage.NO_WORKER_AVAILABLE.getMessage());
  }
  throw new UnavailableException(ExceptionMessage.BLOCK_UNAVAILABLE.getMessage(info.getBlockId()));
  workerPool = locations.stream().map(BlockLocation::getWorkerAddress).collect(toSet());
}
```